pysatMadrigal/instruments/gnss_tec.py

```diff
@@ -0,0 +1,233 @@
+                        "the community by the Massachusetts Institute of ",
+                        "Technology under support from U.S. National Science",
+                        " Foundation grant AGS-1242204. Data for the TEC ",
+                        "processing is provided by the following ",
+                        "organizations: UNAVCO, Scripps Orbit and Permanent",
+                        " Array Center, Institut Geographique National, ",
+                        "France, International GNSS Service, The Crustal ",
+                        "Dynamics Data Information System (CDDIS), National ",
+                        " Geodetic Survey, Instituto Brasileiro de Geografia",
+                        "e Estatística, RAMSAC CORS of Instituto Geográfico",
+                        " Nacional del la República Agentina, Arecibo ",
+                        "Observatory, Low-Latitude Ionospheric Sensor ",
+                        "Network (LISN), Topcon Positioning Systems, Inc., ",
+                        "Canadian High Arctic Ionospheric Network, ",
+                        "Institute of Geology and Geophysics, Chinese ",
+                        "Academy of Sciences, China Meterorology ",
+                        "Administration, Centro di Ricerche Sismogiche, ",
+                        "Système d’Observation du Niveau des Eaux Littorales",
+                        " (SONEL), RENAG : REseau NAtional GPS permanent, ",
+                        "and GeoNet—the official source of geological ",
+                        "hazard information for New Zealand.\n",
+                        mad_meth.cedar_rules()])
+
+    logger.info(ackn_str)
+    self.acknowledgements = ackn_str
+    self.references = "Rideout and Coster (2006) doi:10.1007/s10291-006-0029-5"
+
+    return
+
+
+def download(date_array, tag='', inst_id='', data_path=None, user=None,
+             password=None, url='http://cedar.openmadrigal.org',
+             file_format='netCDF4'):
+    """Downloads data from Madrigal.
+
+    Parameters
+    ----------
+    date_array : array-like
+        list of datetimes to download data for. The sequence of dates need not
+        be contiguous.
+    tag : string
+        Tag identifier used for particular dataset. This input is provided by
+        pysat. (default='')
+    inst_id : string
+        Satellite ID string identifier used for particular dataset. This input
+        is provided by pysat. (default='')
+    data_path : string
+        Path to directory to download data to. (default=None)
+    user : string
+        User string input used for download. Provided by user and passed via
+        pysat. If an account is required for dowloads this routine here must
+        error if user not supplied. (default=None)
+    password : string
+        Password for data download. (default=None)
+    url : string
+        URL for Madrigal site (default='http://cedar.openmadrigal.org')
+    file_format : string
```

Comment:
The `file_format` keyword conflicts with the built-in keyword at the Instrument level. I remember you said something about this during the meeting today but unfortunately I don't remember what. I'm having trouble in this area.

I tried checking out 'hdf5' support. Using the branch over in pysat (without the keyword) I get,
```
gps = pysat.Instrument('gnss', 'tec', tag='vtec')                                                                                                                                                                   
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-8-c7e28318904d> in <module>
----> 1 gps = pysat.Instrument('gnss', 'tec', tag='vtec')

~/Code/pysat/pysat/_instrument.py in _get_supported_keywords(local_func)
   2950         for pop in pop_list[::-1]:
   2951             args.pop(pop)
-> 2952             defaults.pop(pop)
   2953 
   2954     out_dict = {}

IndexError: pop index out of range

```

With the keyword,
```
In [9]: gps = pysat.Instrument('gnss', 'tec', tag='vtec', file_format='hdf5')                                                                                                                                               
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-9-06a989677482> in <module>
----> 1 gps = pysat.Instrument('gnss', 'tec', tag='vtec', file_format='hdf5')

~/Code/pysat/pysat/_instrument.py in __init__(self, platform, name, tag, inst_id, clean_level, update_files, pad, orbit_info, inst_module, multi_file_day, manual_org, directory_format, file_format, temporary_file_list, strict_time_flag, ignore_empty_files, units_label, name_label, notes_label, desc_label, plot_label, axis_label, scale_label, min_label, max_label, fill_label, **kwargs)
    252                 estr = 'file format set to default, supplied string must be '
    253                 estr = '{:s}iteratable [{:}]'.format(estr, self.file_format)
--> 254                 raise ValueError(estr)
    255 
    256         # set up empty data and metadata

ValueError: file format set to default, supplied string must be iteratable [hdf5]
```

Reply:
I was using the wrong pysat branch. You had the correct link above. My bad. Trying again...

Reply:
My test setup was good the first time. Clearly I got myself turned around a bit.

The test for 'hdf5' was done with the 'delimiter_bug` branch. I confirmed that I get the pop issue or the ValueError.

Reply:
Yeah, I apparently changed this bug in the following PR.  Will update it here, too.  Problems of finding too many related problems...